### ISSUE TITLE
RSM with SH?

### ISSUE THREAD

**Issue by Doidel** [mrdoob/three.js#5942]
Hello!

I scanned some indirect illumination methods which are possibly implementable with WebGL and three.js. We have shadow maps, we have deferred rendering, what about implementing reflective shadow maps with SH coefficients? (Here's the RSM paper http://www.vis.uni-stuttgart.de/~dachsbcn/download/rsm.pdf)
Sure, it's only first bounce and has no occlusion, but alteredqualia already once implemented an ssao, and we could also blend it over common shadow maps ofc. Any thoughts/objections?

Negative things are ofc that there's no occlusion involved and that moving the light or something inside the light's frustum makes a reinitialization of the 4 maps necessary (we can ofc still cull the lights though).
Imho it is a feasible simple solution for some first indirection illumination, it's a basis for a more complex GI (Light Propagation Volumes, http://www.vis.uni-stuttgart.de/~dachsbcn/download/lpv.pdf) and chances of success are probably much higher than with e.g. the previously suggested memory heavy Voxel Cone Tracing (https://github.com/mrdoob/three.js/issues/4434 no offense!).

So yeah, I'm looking for some feedback whether it's a good or bad plan!

**Comment by mrdoob** [comment]
This looks like a nice experiment to do, but shouldn't this be better with multiple render targets?

**Comment by Doidel** [comment]
I'm not sure whether I understand your question.
One has to create those 4 maps per light, right? Those can be represented as SH coefficients, but first they ofc have to be rendered. And for that we'd need several FBOs indeed (or compute shaders which we don't really have) which will be turned into SH coefficients. That will cost some time but it's done only once (considering conditions mentioned in first post). If performance is slow there we can still use the draw buffer extension (https://hacks.mozilla.org/2014/01/webgl-deferred-shading/).
After we have those 4 maps as SH coefficients it's normal screen space post processing...
Does this somehow answer your question?

**Comment by mrdoob** [comment]
Yeah, [only 26% of people with WebGL support has `WEBGL_draw_buffers`](http://webglstats.com/) :/

**Comment by Doidel** [comment]
Yeah I know, but this will hardly be the bottleneck, will it? Drawing on 1 texture is really a feasible solution in my eyes. Even if you have 10 lights in your view frustum you'll need to render 40 lowres images when you start up the scene. I don't consider this to e a really annoying thing... much more that it's not made for a very dynamic scene.

**Comment by mrdoob** [comment]
Yep yep! Just wanted to say that this would be a nice experiment like `WebGLDeferredRenderer`. But I wouldn't aim to implement in `WebGLRenderer` directly.

**Comment by arya-s** [comment]
I like this idea, the paper results look promising with what seems little performance hits.

@Doidel are you working on this?

**Comment by Doidel** [comment]
No neither would I @mrdoob 

I sort of disagree @arya-s, It is a GI approach for static lights and static environment. If a light or something in the light's frustum moves a rerender is required (or all 4 maps). In a dynamic scene you'd have quite a performance hit.
Further RSM would be built upon three.js' deferred rendering which already is around 15-20 fps for me, so that bit on top if it might be enough to make it unplayable ^^

I'm not working on this, but I have two students for a bachelor thesis who create something in direction of GI and WebGL. So I'm figuring out what that could be :)

**Comment by Usnul** [comment]
For good deferred rendering you need a way to re-use partial results and combine draw-calls as much as possible. aq implemented deferred renderer in his XG fork of threejs, however that code is not available publicly as far as i'm aware. The point is - his implementation was quite performant, it was built on top of his original work in threejs referred renderer i believe.
